EOS-16512 Implement hare setup.yaml provisioning sections

SwapnilGaonkar7 opened the pull request:
Initial implementation of hare_setup.py to implement steps/sections required by mini-provisioner
TODO: Add rpm version check in post_install stage.

UT: 
Ignore provisioner related error for now.

-bash-4.2$ sudo provisioning/hare_setup.py --post_install
'cortx-s3server' is not installed
Failed to fetch data from provisioner ([Errno 2] No such file or directory: 'provisioner': 'provisioner')
-bash-4.2$ echo $?
0
-bash-4.2$ sudo provisioning/hare_setup.py --init
Cluster is not running
2021-01-25 23:34:12: Generating cluster configuration... OK
2021-01-25 23:34:13: Starting Consul server on this node........... OK
2021-01-25 23:34:22: Importing configuration into the KV store... OK
2021-01-25 23:34:22: Starting Consul on other nodes... OK
2021-01-25 23:34:22: Updating Consul configuraton from the KV store... OK
2021-01-25 23:34:23: Waiting for the RC Leader to get elected....... OK
2021-01-25 23:34:28: Starting Motr (phase1, mkfs)... OK
2021-01-25 23:34:35: Starting Motr (phase1, m0d)... OK
2021-01-25 23:34:48: Starting Motr (phase2, mkfs)... OK
2021-01-25 23:35:11: Starting Motr (phase2, m0d)... OK
2021-01-25 23:35:28: Checking health of services... OK
Stopping m0d@0x7200000000000001:0xc (ios) at ssc-vm-c-0645.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x17 (ios) at ssc-vm-c-0645.colo.seagate.com...
Stopped m0d@0x7200000000000001:0xc (ios) at ssc-vm-c-0645.colo.seagate.com
Stopped m0d@0x7200000000000001:0x17 (ios) at ssc-vm-c-0645.colo.seagate.com
Stopping m0d@0x7200000000000001:0x9 (confd) at ssc-vm-c-0645.colo.seagate.com...
Stopped m0d@0x7200000000000001:0x9 (confd) at ssc-vm-c-0645.colo.seagate.com
Stopping hare-hax at ssc-vm-c-0645.colo.seagate.com...
Stopped hare-hax at ssc-vm-c-0645.colo.seagate.com
Stopping hare-consul-agent at ssc-vm-c-0645.colo.seagate.com...
Stopped hare-consul-agent at ssc-vm-c-0645.colo.seagate.com
Shutting down RC Leader at ssc-vm-c-0645.colo.seagate.com... **ERROR**
Cluster is not running
Failed to fetch data from provisioner ([Errno 2] No such file or directory: 'provisioner': 'provisioner')
-bash-4.2$ echo $?
0
-bash-4.2$ sudo provisioning/hare_setup.py --test
Cluster is not running
2021-01-25 23:49:43: Generating cluster configuration... OK
2021-01-25 23:49:44: Starting Consul server on this node......... OK
2021-01-25 23:49:50: Importing configuration into the KV store... OK
2021-01-25 23:49:51: Starting Consul on other nodes... OK
2021-01-25 23:49:51: Updating Consul configuraton from the KV store... OK
2021-01-25 23:49:52: Waiting for the RC Leader to get elected....... OK
2021-01-25 23:49:56: Starting Motr (phase1, mkfs)... OK
2021-01-25 23:50:02: Starting Motr (phase1, m0d)... OK
2021-01-25 23:50:16: Starting Motr (phase2, mkfs)... OK
2021-01-25 23:50:39: Starting Motr (phase2, m0d)... OK
2021-01-25 23:50:56: Checking health of services... OK
Stopping m0d@0x7200000000000001:0xc (ios) at ssc-vm-c-0645.colo.seagate.com...
Stopping m0d@0x7200000000000001:0x17 (ios) at ssc-vm-c-0645.colo.seagate.com...
Stopped m0d@0x7200000000000001:0xc (ios) at ssc-vm-c-0645.colo.seagate.com
Stopped m0d@0x7200000000000001:0x17 (ios) at ssc-vm-c-0645.colo.seagate.com
Stopping m0d@0x7200000000000001:0x9 (confd) at ssc-vm-c-0645.colo.seagate.com...
Stopped m0d@0x7200000000000001:0x9 (confd) at ssc-vm-c-0645.colo.seagate.com
Stopping hare-hax at ssc-vm-c-0645.colo.seagate.com...
Stopped hare-hax at ssc-vm-c-0645.colo.seagate.com
Stopping hare-consul-agent at ssc-vm-c-0645.colo.seagate.com...
Stopped hare-consul-agent at ssc-vm-c-0645.colo.seagate.com
Shutting down RC Leader at ssc-vm-c-0645.colo.seagate.com... **ERROR**
Cluster is not running
Failed to fetch data from provisioner ([Errno 2] No such file or directory: 'provisioner': 'provisioner')
-bash-4.2$ echo $?
0

-bash-4.2$ sudo provisioning/hare_setup.py --reset
-bash-4.2$ echo $?
0
-bash-4.2$ sudo provisioning/hare_setup.py --cleanup
-bash-4.2$ echo $?
0

-bash-4.2$ sudo provisioning/hare_setup.py --support_bundle
Cluster is not running
Created /tmp/hare/hare_ssc-vm-c-0645.tar.gz
Failed to fetch data from provisioner ([Errno 2] No such file or directory: 'provisioner': 'provisioner')

SwapnilGaonkar7 commented:
@mssawant @knekrasov @s-arya Please review

knekrasov commented on provisioning/setup.py:
```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
+            nodes_data = cluster_info['nodes']
+
+            for nodeinfo in nodes_data:
+                if nodeinfo[0] == hostname:
```
@SwapnilGaonkar7 I'm afraid you're relying on a broken JSON format. 

It is fixed by https://github.com/Seagate/cortx-hare/pull/1397
@mssawant would you mind to review it finally? 😄

mssawant replied:
Merged!

knekrasov replied:
@SwapnilGaonkar7 please merge the freshest `main` branch into your PR and accommodate the changes in `hctl status --json` output (an example can be also seen in #1397 description). Thanks!

SwapnilGaonkar7 replied:
Done

knekrasov reviewed (commented):
There are some things to discuss. I'd wait for @mssawant 's comments.

mssawant commented on provisioning/setup.py:
```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
                    nargs=0,
                    help='Report unsupported features according to setup type',
                    action=UnsupportedFeatures)
+    p.add_argument('--post_install',
```
I did think that it would be good to merge setup.py with hax regular sources because it is possible provisioning/setup.py may grow further. Testing can be done by running setup.py utility instead of functions, but it makes sense to modularize it but I think we can take it up as a separate task.

knekrasov replied:
Created #1464. No need to handle this suggestion in scope of this PR.

mssawant commented on provisioning/setup.py:
```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
+    cluster_desc = None
+    with open("/var/lib/hare/cluster.yaml", 'r') as stream:
+        cluster_desc = yaml.safe_load(stream)
+    hostname = socket.gethostname()
```
Why do we need to get the current hostname? if we are checking the entire cluster setup the algorithm should be,
- get cluster status `hctl status --json`
- load CDF and iterate over all the nodes,
  - if node runs_confd from CDF, check if corresponding node (using hostname) is running confd and if it is `started` in the dictionary loaded from `hctl status --json` output.
  - check for ioservices for each node.
  
something like,

def check_node_status(node_data_cdf: dict, node_data_hctl: dict) -> bool:
... 

def check_cluster_status() -> bool:
//load cdf
//load `hctl status --json` output
     for node_cdf in cluster_desc['nodes']:
         node_data_cdf = node.get('m0_servers', [])
         node_data_hctl = cluster_info['nodes'][node_data_cdf.hostname]
         if not check_node_status(node_data_cdf, node_data_hctl):
             logging.error()
             return False

Makes sense?

mssawant replied:
@SwapnilGaonkar7 try this,
```
def list2dict(nodes_data_hctl: List) -> dict:
    node_info_dict = {}
    for node in nodes_data_hctl:
        node_svc_info = {}
        for service in node[1]:
            node_svc_info[service[0]] = []
            if (service[3] == 'started'):
                node_svc_info[service[0]].append(service[3])
        node_info_dict[node[0]] = node_svc_info

    return node_info_dict

def check_cluster_status_new() -> bool:
    cluster_desc = None
    with open("/tmp/cluster-6-1-ios.yaml", 'r') as stream:
        cluster_desc = yaml.safe_load(stream)
    cmd = 'hctl status --json'
    cluster_info = json.loads(execute_command_and_get_output(cmd))
    nodes_data_hctl = cluster_info['nodes']

    # node_info_dict = {node[0]: {service[0]: append(service[3]) for service in node[1]} for node in nodes_data_hctl}
    #node_info_dict = {node[0]: {}.setdefault(service[0], []).append(service[3]) for service in node[1] for node in nodes_data_hctl}
    node_info_dict = list2dict(nodes_data_hctl)
    # print(node_info_dict)
    for node in cluster_desc['nodes']:
        m0client_s3_cnt = node['m0_clients']['s3']
        m0client_other_cnt = node['m0_clients']['other']
        m0ds = node.get('m0_servers', [])
        for m0d in m0ds:
            ios_cnt = 0
            if ('runs_confd' in m0d.keys() and
                (node_info_dict[node['hostname']]['confd'][0] != 'started')):
                print('returning false for confd ' + node_info_dict[node['hostname']]['confd'][0])
                return False
            if m0d['io_disks']['data']:
                if node_info_dict[node['hostname']]['ioservice'][ios_cnt] != 'started':
                    print('returning false for ios ' + node['hostname'])
                    return False
                ios_cnt += 1
        if len(node_info_dict[node['hostname']]['s3server']) != m0client_s3_cnt:
            print('returning false for s3servers ' + node['hostname'])
            return False

    return True
```
I specifically omitted m0_clients status check, although its included in the dictionary because they are optional processes.

mssawant commented:
please rename provisioning/setup.py to provisioning/hare_setup.py to avoid confusion between python setup.py.

SwapnilGaonkar7 commented on provisioning/setup.py:
```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
+class Test(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            if check_cluster_status() != 0:
```
@mssawant If cluster is not running and we do bootstrap then we should call shutdown after 'hctl status' right?

mssawant replied:
Yes, test or init, both should come out with a stopped cluster.

knekrasov commented on provisioning/setup.py:
```diff
@@ -83,14 +101,214 @@ def __call__(self, parser, namespace, values, option_string=None):
+                          error)
+            exit(-1)
+
+def checkRpm(rpm_name):
```
I would vote against generic validators.
1. I find their design is overcomplicated without actual benefit of that complexity (please see more details in [my ticket](https://jts.seagate.com/browse/EOS-14723?focusedCommentId=1990763&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1990763)).
2. We can't control that code even if we're the authors. Code review there tends to lower code quality.

Let's keep the code in our repository as long as we can.
cc @NeeravChoudhari

SwapnilGaonkar7 commented:
@mssawant @knekrasov Please have a look. I have addressed review comments

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+    node_info_dict = list2dict(nodes_data_hctl)
+    for node in cluster_desc['nodes']:
+        m0client_s3_cnt = node['m0_clients']['s3']
+        m0client_other_cnt = node['m0_clients']['other']
```
Please remove this as we are not using it.

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+    return 0
+
+
+def check_cluster_status_old():
```
Please remove this if not required.

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+            if checkRpm('cortx-motr') != 0:
+                logging.error('\'cortx-motr\' is not installed')
+                exit(-1)
+            if checkRpm('consul') != 0:
```
let's add a consul version check to match what we have in our hare.spec.

SwapnilGaonkar7 replied:
@mssawant I was planning to it under separate subtask.

mssawant replied:
okay.

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+                if node_info_dict[node['hostname']]['ioservice'][ios_cnt] != 'started':
+                  return -1
+                ios_cnt += 1
+        if ('s3server' in node_info_dict[node['hostname']].keys() and
```
You don't need to check if 's3server' exists in the keys, if s3server count in CDF is 0 then `node_info_dict[node['hostname']]['s3server']` status list will be empty.

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE,
+                                        shell=True)
+        stdout, stderr = process.communicate()
```
stdout, _ = process.communicate() ?

knekrasov replied:
Exit code is not checked. So if the invoked command returns non-zero, the problem can be swallowed by mistake.

Please have a look how it works in pcswrap: https://github.com/knekrasov/cortx-ha/blob/main/ha/pcswrap/pcswrap/internal/connector.py#L72 
Probably you can follow the same idea.

SwapnilGaonkar7 replied:
Done

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+class Cleanup(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            logging.info('No-op for Cleanup/Reset as of now for Hare ')
```
I don't think this log message is needed, it does not provide any consumable information as such.

mssawant replied:
@SwapnilGaonkar7, I think you removed `Cleanup` class altogether, we need to support the cleanup function, just that it can be a no-op for now. In the above comment I just asked you to remove the log message as it did not provide much information.

SwapnilGaonkar7 replied:
Done

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+
+
+def is_cluster_running():
+    if os.system('hctl status >/dev/null') == 0:
```
Please
```
def is_cluster_running() -> bool:
    return os.system('hctl status >/dev/null') == 0
```

SwapnilGaonkar7 replied:
Done

SwapnilGaonkar7 replied:
Done

knekrasov replied:
Hm.. have you pushed your latest changes @SwapnilGaonkar7 ? 😄

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+        try:
+            cluster_started = False
+            cluster_status_failure = False
+            if is_cluster_running() == False:
```
```
if not is_cluster_runing() and bootstrap_cluster() != 0:
    logging.error('Failed to bootstrap the cluster')
    exit(-1)
```

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+class Init(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            if is_cluster_running() == True:
```
I think it should be,
```
if not is_cluster_running() and bootstrap_cluster() != 0:
    logging.error('Failed to bootstrap the custer')
    exit(1)

while is_cluster_running():
    shutdown_cluster()
```

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+
+def get_data_from_provisioner_cli(method, output_format='json') -> str:
+    try:
+        process = subprocess.run(['provisioner', method, f'--out={output_format}'],
```
Why don't you add a helper function that invokes the given command properly (like [`_execute()`](https://github.com/knekrasov/cortx-ha/blob/main/ha/pcswrap/pcswrap/internal/connector.py#L72) in my previous example from pcswrap).

That will make the code shorter and will hide the ugly details of `subprocess.run()`. And Codacy findings can be addressed there, within a single place :wink:

SwapnilGaonkar7 replied:
@knekrasov This is existing code showing here because of file rename. Can I have handle it in separate task as provisioner setup will be required to test the changes

knekrasov replied:
Well, I think that provisioner setup just to ensure that the code is not broken is an overkill, but OK, let's  defer this refactoring. Please ignore the comment and concentrate on your changes.

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,383 @@
+            exit(-1)
+
+def checkRpm(rpm_name):
+    return os.system('rpm -qa | grep -q ' + rpm_name)
```
@SwapnilGaonkar7 If you're curious, there is a way to invoke piped commands using `subprocess`. Probably that will make Codacy happier.

Here is the example how it can be done in our Hare commit history: https://github.com/Seagate/cortx-hare/blob/4e360eec9b5501d26af7ded61ad139691283c8e6/ha-notifier/bq/listener.py#L39

There, I invoke `m0hagen | m0ham` through this api.

SwapnilGaonkar7 replied:
Done

SwapnilGaonkar7 commented:
@knekrasov @mssawant Please have a look

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,279 @@
+
+from cortx.utils.product_features import unsupported_features
+
+def execute_command_and_get_output(cmd):
```
`def execute(cmd: str) -> Tuple[str, int]` is enough.

Btw do you really need to return exit code? The thing is that normally I don't care what was the exit code. Let it just raise an exception if exit code was not 0. So the caller doesn't need to re-check exit code. If exit code is not good, there will be an exception (so the error is not swallowed), but if the things are fine (exit code = 0), then I don't need to analyze it specially. I'll just concentrate on positive scenario in the code.

SwapnilGaonkar7 replied:
DOone

mssawant replied:
@SwapnilGaonkar7, you its done but somehow I don't see the updated code. Also please check the codacy warning about `returncode`.
As the matter of fact please see if you can address the rest of the codacy warnings/errors too.
Thank you.

SwapnilGaonkar7 replied:
@mssawant @knekrasov I have fixed above issue and also fixed few codacy warnings/errors.
I have also done some modification in 'checkRpm' function.
Please have a look

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,279 @@
+    cluster_info_out, err = execute_command_and_get_output(cmd)
+    if err:
+        logging.error('Failed to get hctl status')
+        return -1
```
What is the point in returning a magic number? If there is a problem - please raise an exception.

SwapnilGaonkar7 replied:
Done

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,279 @@
+    return 0
+
+
+def main(argv=None):
```
I guess `argv` are always None. Probably you should remove them from the signature.

SwapnilGaonkar7 replied:
Done

SwapnilGaonkar7 commented:
@knekrasov Please have a look

mssawant commented:
@SwapnilGaonkar7 , can you post the testing output with the updated patch please?

SwapnilGaonkar7 commented:
@mssawant @knekrasov I have addressed all the review comments until now and also commented as done on them.
If I miss anything please let me know
Also updated description with testing output of latest code

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,305 @@
+                                        encoding='utf8')
+        out, _ = process.communicate()
+        if process.returncode:
+            raise Exception('Error occured while executing command')
```
Usually when an error happens, the details can be found in stderr. Can we add stderr to the exception?

like `raise Exception(f'Command {cmd} exited with error code {returncode}. Command output: {err}')

knekrasov replied:
...or at least log the error details with `logger.error()`. Just imagine, how would you debug an issue that happened in the field and how much of information would be sufficient to understand what was wrong.

SwapnilGaonkar7 replied:
Done

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,305 @@
+
+
+def is_cluster_running() -> bool:
+    if os.system('hctl status >/dev/null') == 0:
```
```
if <bool>:
   return <bool>
else
   return <bool>
```

looks really weird. Please consider replacing it with something like 
```python
return os.system('hctl status >/dev/null') == 0
```

SwapnilGaonkar7 replied:
Done

SwapnilGaonkar7 commented:
@knekrasov Please have a look

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,301 @@
+    with open("/var/lib/hare/cluster.yaml", 'r') as stream:
+        cluster_desc = yaml.safe_load(stream)
+    cmd = ['hctl', 'status', '--json']
+    cluster_info = json.loads(execute(cmd))
```
`execute` consumes `cmd: str` but instead `List[str]` is passed

SwapnilGaonkar7 replied:
Done

knekrasov commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,301 @@
+        for m0d in m0ds:
+            ios_cnt = 0
+            if ('runs_confd' in m0d.keys() and
+                (node_info_dict[node['hostname']]['confd'][0] != 'started')):
```
Does this code work with the JSON structure like this?

```json
 "nodes": [
    {
      "name": "localhost",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x6",
          "ep": "192.168.6.214@tcp:12345:1:1",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x9",
          "ep": "192.168.6.214@tcp:12345:2:1",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0xc",
          "ep": "192.168.6.214@tcp:12345:2:2",
          "status": "started"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x28",
          "ep": "192.168.6.214@tcp:12345:4:1",
          "status": "unknown"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x2b",
          "ep": "192.168.6.214@tcp:12345:4:2",
          "status": "unknown"
        }
      ]
    }
  ]
```
If yes, I don't quite understand why there is 0 index in your expression `node_info_dict[node['hostname']]['confd'][0] != 'started'`

SwapnilGaonkar7 replied:
node_info_dict[node['hostname']]['confd'] is a list. Actually node_info_dict[node['hostname']][<service name>] is list and in case of 'ioservices' it will contain multiple entries of status(started).
In case of confd only 1 entry will be there thats why node_info_dict[node['hostname']]['confd'][0] != 'started'

mssawant commented on provisioning/hare_setup.py:
```diff
@@ -0,0 +1,303 @@
+
+def check_cluster_status():
+    cluster_desc = None
+    with open("/var/lib/hare/cluster.yaml", 'r') as stream:
```
I don't think it is a good idea to hardcode the cluster.yaml path. I think it would be good to accept it as an argument, either via conf-store or an explicit argument to hare-setup.py. But let's postpone this for now and we improve it later.

mssawant reviewed (approved):
Done with the review round. Further improvements can be done as part of the separate tasks.
